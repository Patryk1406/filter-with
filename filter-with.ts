import { isDeepStrictEqual } from 'util';

const isString = (value: unknown) => typeof value === 'string';

const hasThreeConsecutiveChars = (value: string) => /\w{3,}/.test(value.trim());

const canBeSearchedString = (value: unknown): boolean => isString(value)
  && !hasThreeConsecutiveChars(value as string);

const isSetOrMap = (value: unknown): boolean => value instanceof Set || value instanceof Map;

const isFunction = (value: unknown): boolean => typeof value === 'function';

const isEmptyArray = (value: unknown): boolean => Array.isArray(value) && value.length === 0;

const cannotBeTarget = (value: unknown): boolean => !value || isFunction(value)
  || canBeSearchedString(value)
  || isEmptyArray(value);

const searchThroughObject = <T, K>(
  obj: T, target: K): boolean => Object.values(obj)
    .some((value) => {
      if (cannotBeTarget(value)) {
        return false;
      }
      if (isDeepStrictEqual(value, target)) return true;
      if (typeof value === 'object') {
        return isSetOrMap(value)
          ? searchThroughObject(Array.from(value.values()), target)
          : searchThroughObject(value, target);
      }
      return false;
    });

export const filterWith = <T, K>(
  array: Record<string, T>[], elementToFind: K): [] | Record<string, T>[] => {
  if (!elementToFind
    || canBeSearchedString(elementToFind)
    || array.length === 0) {
    return [];
  }
  return array.filter((element) => searchThroughObject(element, elementToFind));
};
