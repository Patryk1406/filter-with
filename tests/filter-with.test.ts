import { filterWith } from '../filter-with';

describe('FilterWith function test suite', () => {
  const inputData = [
    {
      id: '5e985a07feddae7617ac44f6',
      age: 24,
      eyeColor: 'brown',
      name: 'Cummings Baxter',
      gender: 'male',
      company: 'VELOS',
      email: 'cummingsbaxter@velos.com',
      phone: '+1 (907) 482-2451',
      tags: ['labore', 'elit', 'excepteur', 'nisi', 'mollit', 'anim', 'aliquip'],
      friends: [
        {
          id: 0,
          name: 'Sheppard Jensen',
        },
      ],
    },
    {
      id: '5e985a0709dfa1e6fd93c6ad',
      age: 32,
      eyeColor: 'brown',
      name: 'Madelyn Dickson',
      gender: 'female',
      company: 'KENGEN',
      email: 'madelyndickson@kengen.com',
      phone: '+1 (984) 521-2439',
      tags: ['nisi', 'veniam', 'dolore', 'officia', 'ex', 'non', 'pariatur'],
      friends: [
        {
          id: 0,
          name: 'Bruce Barton',
        },
        {
          id: 1,
          name: 'Juliet Schmidt',
        },
        {
          id: 2,
          name: 'Horton Haley',
        },
        {
          id: 3,
          name: 'Herminia Witt',
        },
      ],
    },
    {
      id: '5e985a0737e2306e9aef6ecd',
      age: 26,
      eyeColor: 'blue',
      name: 'Mcguire Mercado',
      gender: 'male',
      company: 'LINGOAGE',
      email: 'mcguiremercado@lingoage.com',
      phone: '+1 (963) 450-2194',
      tags: ['cupidatat', 'occaecat', 'amet', 'qui', 'elit', 'esse', 'deserunt'],
      friends: [
        {
          id: 0,
          name: 'Loraine Harper',
        },
        {
          id: 1,
          name: 'Luann Randall',
        },
        {
          id: 2,
          name: 'Obrien Rich',
        },
        {
          id: 3,
          name: 'Noble Wilkerson',
        },
      ],
    },
    {
      id: '5e985a07148cfba58c860ec2',
      age: 266,
      eyeColor: 'brown',
      name: 'Marina Porter',
      gender: 'female',
      company: 'GORGANIC',
      email: 'marinaporter@gorganic.com',
      phone: '+1 (867) 417-3497',
      tags: ['laborum', 'aliquip', 'sit', 'adipisicing', 'aute', 'cupidatat', 'aliquip'],
      friends: [
        {
          id: 0,
          name: 'Blair Hill',
        },
        {
          id: 123,
          name: 'Ebony Jimenez',
        },
      ],
    },
    {
      id: '5e985a074984f9f08ccaaa4c',
      age: 25,
      eyeColor: 'green',
      name: 'Barlow Ferguson',
      gender: 'male',
      company: 'TOYLETRY',
      email: 'barlowferguson@toyletry.com',
      phone: '+1 (837) 484-2231',
      tags: ['est', 'dolor', 'minim', 'ut', 'anim', 'culpa', 'non'],
      friends: [
        {
          id: 0,
          name: 'Delacruz Acevedo',
        },
        {
          id: 1,
          name: 'Gloria Tanner',
        },
        {
          id: 2,
          name: 'Cantrell Myers',
        },
        {
          id: 3,
          name: 'Fisher Leonard',
        },
      ],
    },
  ];

  test('Filter an array by a not embedded string field.', () => {
    const elementToFind = 'Cummings Baxter';

    const output = filterWith(inputData, elementToFind);

    expect(output)
      .toMatchSnapshot();
  });

  test('Filter an array by an embedded string field in the tags object.', () => {
    const elementToFind = 'nisi';

    const output = filterWith(inputData, elementToFind);

    expect(filterWith(output, 'nisi'))
      .toMatchSnapshot();
  });

  test('Filter an array by an embedded string field in the friends object.', () => {
    const elementToFind = 'Delacruz Acevedo';

    const output = filterWith(inputData, elementToFind);

    expect(output)
      .toMatchSnapshot();
  });

  test(
    'Return an empty array when the phrase has less than consecutive 3 chars or an empty array is given.',
    () => {
      const elementToFind = 'L ul';

      const output = filterWith(inputData, elementToFind);

      const expectedOutput: Record<string, string>[] = [];

      expect(output)
        .toStrictEqual(expectedOutput);

      const emptyArrayInput: Record<string, string>[] = [];

      const elementToFind2 = 266;

      const output2 = filterWith(emptyArrayInput, elementToFind2);

      expect(output2)
        .toStrictEqual(expectedOutput);
    },
  );

  test('Filter an array by a not embedded number field', () => {
    const elementToFind = 266;

    const output = filterWith(inputData, elementToFind);

    expect(output)
      .toMatchSnapshot();
  });

  test('Filter an array by an embedded number field in the friends object.', () => {
    const elementToFind = 266;

    const output = filterWith(inputData, elementToFind);

    expect(output)
      .toMatchSnapshot();
  });

  test('Filter an array by an array', () => {
    const inputArray: Record<string, string | number | string[]>[] = [
      {
        string: 'abc',
        arr: ['a', 'b'],
      },
      {
        number: 34,
        string: 'www',
        set: 4,
      }];

    const elementToFind = ['a', 'b'];

    const output = filterWith(inputArray, elementToFind);

    expect(output)
      .toMatchSnapshot();
  });

  test('Filter an array by a set', () => {
    const inputArray: Record<string, string | number | Set<number> | Set<Set<number>>>[] = [
      {
        string: 'abc',
        set: new Set([new Set([1, 2, 3, 4])]),
      },
      {
        number: 34,
        string: 'www',
        set: new Set([1]),
      }];

    const elementToFind = new Set([1, 2, 3, 4]);

    const output = filterWith(inputArray, elementToFind);

    expect(output)
      .toMatchSnapshot();
  });

  test('Filter an array by a number embedded in a map', () => {
    const inputArray: Record<string, string | number | Set<Set<number>> | Map<string, number>>[] = [
      {
        string: 'abc',
        set: new Set([new Set([2, 3, 4])]),
      },
      {
        number: 34,
        string: 'www',
        set: new Map([['a', 1], ['b', 2]]),
      }];

    const elementToFind = 1;

    const output = filterWith(inputArray, elementToFind);

    expect(output)
      .toMatchSnapshot();
  });

  test('Filter an array by a map', () => {
    const inputArray: Record<string, string | number | Set<Set<number>> | Map<string, number>>[] = [
      {
        string: 'abc',
        set: new Set([new Set([2, 3, 4])]),
      },
      {
        number: 34,
        string: 'www',
        set: new Map([['a', 1], ['b', 2]]),
      }];

    const elementToFind = new Map([['a', 1], ['b', 2]]);

    const output = filterWith(inputArray, elementToFind);

    expect(output)
      .toMatchSnapshot();
  });
});
