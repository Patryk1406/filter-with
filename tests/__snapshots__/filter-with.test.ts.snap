// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterWith function test suite Filter an array by a map 1`] = `
Array [
  Object {
    "number": 34,
    "set": Map {
      "a" => 1,
      "b" => 2,
    },
    "string": "www",
  },
]
`;

exports[`FilterWith function test suite Filter an array by a not embedded number field 1`] = `
Array [
  Object {
    "age": 266,
    "company": "GORGANIC",
    "email": "marinaporter@gorganic.com",
    "eyeColor": "brown",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Blair Hill",
      },
      Object {
        "id": 123,
        "name": "Ebony Jimenez",
      },
    ],
    "gender": "female",
    "id": "5e985a07148cfba58c860ec2",
    "name": "Marina Porter",
    "phone": "+1 (867) 417-3497",
    "tags": Array [
      "laborum",
      "aliquip",
      "sit",
      "adipisicing",
      "aute",
      "cupidatat",
      "aliquip",
    ],
  },
]
`;

exports[`FilterWith function test suite Filter an array by a not embedded string field. 1`] = `
Array [
  Object {
    "age": 24,
    "company": "VELOS",
    "email": "cummingsbaxter@velos.com",
    "eyeColor": "brown",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Sheppard Jensen",
      },
    ],
    "gender": "male",
    "id": "5e985a07feddae7617ac44f6",
    "name": "Cummings Baxter",
    "phone": "+1 (907) 482-2451",
    "tags": Array [
      "labore",
      "elit",
      "excepteur",
      "nisi",
      "mollit",
      "anim",
      "aliquip",
    ],
  },
]
`;

exports[`FilterWith function test suite Filter an array by a number embedded in a map 1`] = `
Array [
  Object {
    "number": 34,
    "set": Map {
      "a" => 1,
      "b" => 2,
    },
    "string": "www",
  },
]
`;

exports[`FilterWith function test suite Filter an array by a set 1`] = `
Array [
  Object {
    "set": Set {
      Set {
        1,
        2,
        3,
        4,
      },
    },
    "string": "abc",
  },
]
`;

exports[`FilterWith function test suite Filter an array by an array 1`] = `
Array [
  Object {
    "arr": Array [
      "a",
      "b",
    ],
    "string": "abc",
  },
]
`;

exports[`FilterWith function test suite Filter an array by an embedded number field in the friends object. 1`] = `
Array [
  Object {
    "age": 266,
    "company": "GORGANIC",
    "email": "marinaporter@gorganic.com",
    "eyeColor": "brown",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Blair Hill",
      },
      Object {
        "id": 123,
        "name": "Ebony Jimenez",
      },
    ],
    "gender": "female",
    "id": "5e985a07148cfba58c860ec2",
    "name": "Marina Porter",
    "phone": "+1 (867) 417-3497",
    "tags": Array [
      "laborum",
      "aliquip",
      "sit",
      "adipisicing",
      "aute",
      "cupidatat",
      "aliquip",
    ],
  },
]
`;

exports[`FilterWith function test suite Filter an array by an embedded string field in the friends object. 1`] = `
Array [
  Object {
    "age": 25,
    "company": "TOYLETRY",
    "email": "barlowferguson@toyletry.com",
    "eyeColor": "green",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Delacruz Acevedo",
      },
      Object {
        "id": 1,
        "name": "Gloria Tanner",
      },
      Object {
        "id": 2,
        "name": "Cantrell Myers",
      },
      Object {
        "id": 3,
        "name": "Fisher Leonard",
      },
    ],
    "gender": "male",
    "id": "5e985a074984f9f08ccaaa4c",
    "name": "Barlow Ferguson",
    "phone": "+1 (837) 484-2231",
    "tags": Array [
      "est",
      "dolor",
      "minim",
      "ut",
      "anim",
      "culpa",
      "non",
    ],
  },
]
`;

exports[`FilterWith function test suite Filter an array by an embedded string field in the tags object. 1`] = `
Array [
  Object {
    "age": 24,
    "company": "VELOS",
    "email": "cummingsbaxter@velos.com",
    "eyeColor": "brown",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Sheppard Jensen",
      },
    ],
    "gender": "male",
    "id": "5e985a07feddae7617ac44f6",
    "name": "Cummings Baxter",
    "phone": "+1 (907) 482-2451",
    "tags": Array [
      "labore",
      "elit",
      "excepteur",
      "nisi",
      "mollit",
      "anim",
      "aliquip",
    ],
  },
  Object {
    "age": 32,
    "company": "KENGEN",
    "email": "madelyndickson@kengen.com",
    "eyeColor": "brown",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Bruce Barton",
      },
      Object {
        "id": 1,
        "name": "Juliet Schmidt",
      },
      Object {
        "id": 2,
        "name": "Horton Haley",
      },
      Object {
        "id": 3,
        "name": "Herminia Witt",
      },
    ],
    "gender": "female",
    "id": "5e985a0709dfa1e6fd93c6ad",
    "name": "Madelyn Dickson",
    "phone": "+1 (984) 521-2439",
    "tags": Array [
      "nisi",
      "veniam",
      "dolore",
      "officia",
      "ex",
      "non",
      "pariatur",
    ],
  },
]
`;
